app/controllers/uploads_controller.rb

```diff
@@ -110,6 +110,17 @@ def show_short
     end
   end
 
+  def show_secure
+    # do not serve uploads requested via XHR to prevent XSS
+    return xhr_not_allowed if request.xhr?
+
+    if SiteSetting.secure_media?
+      redirect_to Discourse.store.signed_url_for_path("#{params[:path]}.#{params[:extension]}")
```

Comment:
Could this be a DoS vector? Should this be rate limited and/or cached?

Reply:
I don't think so because the `show_secure` method requires a logged in user.